From ea5f9f15329d8a2ffdf98db6d2b8c2210a1d0260 Mon Sep 17 00:00:00 2001
From: Eric Gillespie <epg@pretzelnet.org>
Date: Thu, 11 Dec 2025 17:46:16 -0600
Subject: [PATCH 37/74] Reuse exit code of  'create --command' command

When the child exits with non-zero status, use it as snapper's own exit
code, as it may be meaningful.

We pass the exit code up the stack with a new CommandException type.
---
 client/snapper/cmd-create.cc |  4 ++--
 client/snapper/cmd.h         | 10 ++++++++++
 client/snapper/snapper.cc    |  5 +++++
 doc/snapper.xml.in           |  4 ++--
 4 files changed, 19 insertions(+), 4 deletions(-)

diff --git a/client/snapper/cmd-create.cc b/client/snapper/cmd-create.cc
index 65cbaeec..477b711b 100644
--- a/client/snapper/cmd-create.cc
+++ b/client/snapper/cmd-create.cc
@@ -25,6 +25,7 @@
 
 #include <iostream>
 
+#include <client/snapper/cmd.h>
 #include <snapper/AppUtil.h>
 
 #include "../utils/text.h"
@@ -205,8 +206,7 @@ namespace snapper
 		if (WIFEXITED(status)) {
 		    int const exit_status = WEXITSTATUS(status);
 		    if (exit_status != 0) {
-			// TODO(epg): Can we have snapper itself exit with this status?
-			SN_THROW(Exception(sformat("%s exited %d", command.c_str(), exit_status)));
+			SN_THROW(CommandException(exit_status));
 		    }
 		} else {
 		    SN_THROW(Exception(sformat("%s killed with %d", command.c_str(), WTERMSIG(status))));
diff --git a/client/snapper/cmd.h b/client/snapper/cmd.h
index 7bb5c423..e55c08f4 100644
--- a/client/snapper/cmd.h
+++ b/client/snapper/cmd.h
@@ -195,4 +195,14 @@ namespace snapper
     command_debug(GlobalOptions& global_options, GetOpts& get_opts, ProxySnappers* snappers,
 		  ProxySnapper* snapper, Plugins::Report& report);
 
+    struct CommandException : public Exception
+    {
+	explicit CommandException(int exit_status)
+	    : Exception(),
+	      exit_status(exit_status)
+	{}
+
+	const int exit_status;
+    };
+
 }
diff --git a/client/snapper/snapper.cc b/client/snapper/snapper.cc
index e2ce27ab..a161f494 100644
--- a/client/snapper/snapper.cc
+++ b/client/snapper/snapper.cc
@@ -368,6 +368,11 @@ main(int argc, char** argv)
 	    cerr << e.what() << endl;
 	    exit(EXIT_FAILURE);
 	}
+	catch (const CommandException& e)
+	{
+	    SN_CAUGHT(e);
+	    exit(e.exit_status);
+	}
 	catch (const Exception& e)
 	{
 	    SN_CAUGHT(e);
diff --git a/doc/snapper.xml.in b/doc/snapper.xml.in
index 2b4a793c..296ddfb9 100644
--- a/doc/snapper.xml.in
+++ b/doc/snapper.xml.in
@@ -537,8 +537,8 @@
 	      <listitem>
 		<para>Create a pre and post snapshot and run command in between.  If snapper is
 		unable to execute the command at all, it does not take the post snapshot.  If the
-		command exits with a non-zero exit code, snapper takes the post snapshot and exits 1
-		to indicate the failure.</para>
+		command exits with a non-zero exit code, snapper takes the post snapshot and exits
+		with that same exit code.</para>
 	      </listitem>
 	    </varlistentry>
 	    <varlistentry>
-- 
2.43.0

